microprofile/WFLY-12313_MicroProfile_OpenAPI_1.1.adoc

```diff
@@ -0,0 +1,111 @@
+= Integrate MicroProfile OpenAPI 1.1
+:author:            Paul Ferraro
+:email:             paul.ferraro@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+This enhancement proposes to add support for https://download.eclipse.org/microprofile/microprofile-open-api-1.1.2/microprofile-openapi-spec.html[Eclipse MicroProfile OpenAPI 1.1] to WildFly by integrating the implementation from https://github.com/smallrye/smallrye-open-api[SmallRye].
+This spec provides a unified Java API for producing documentation that complies with the https://github.com/OAI/OpenAPI-Specification/blob/3.0.2/versions/3.0.2.md[OpenAPI v3 specification].
+
+An MP OpenAPI subsystem exposes an HTTP GET `/openapi` endpoint that returns an OpenAPI v3 compliant document (in YAML or JSON format) containing documentation of the RESTful services of an application.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFLY-12313
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:fburzigo@redhat.com[Fabio Burzigotti]
+
+=== Testing By
+
+[ ] Engineering
+
+[X] QE
+
+=== Affected Projects or Components
+
+The SmallRye implementation of the MicroProfile OpenAPI specification is already integrated by the following projects:
+
+* https://thorntail.io/[Thorntail] via the https://github.com/thorntail/thorntail/tree/master/fractions/microprofile/microprofile-openapi[microprofile-openapi fraction].
+* https://quarkus.io/[Quarkus] via the https://github.com/quarkusio/quarkus/tree/master/extensions/smallrye-openapi[smallrye-openapi extension].
+
+These integrations will heavily inform, though not dictate, the corresponding integration into WildFly.
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+* Create an MP OpenAPI galleon layer
+** Should depend on MP Config and JAX-RS (and their dependencies)
+* The microprofile-openapi spec module must be exposed to deployments
+* The `/openapi` endpoint shall be registered relative to the root of the virtual host associated with the deployment.
+** This places a superficial constraint of a single application per host (see subsequent section re: multiple deployments)
+** In the event a web deployment uses a given host for which another deployment has already registered an `/openapi` endpoint, we will:
+... Log a warning.
+... Skip OpenAPI model generation for that deployment.
+* All WildFly/SmallRye specific configuration will be obtained via MP Config using the properties prefixed with `mp.openapi.extensions`.
```

Comment:
Technically, such an attribute is redundant with the /subsystem=microprofile-openapi:add()/remove() operations, but using attribute makes it easier to parameterize this behavior for a given environment.

Reply:
Ok, if you agree, let's leave the subsystem activation/deactivation via the CLI, which is not relevant here since we're actually discussing (extended) MP config examples.
That said, IIUC we can state there will be one only _extended_ configuration feature, which the user will be able to use, i.e. setting one or more properties, in order to 

> enable/disable /openapi registration

(see your comment [1])

Right? If the above is correct, then I think the properties/property name/s should be defined in the document and the above mentioned case could be used as the example we're discussing about. WDYT?  

One more question: when you state:

> ... parameterize this behavior for a given environment

Do you mean "_for a given application and thus not globally_"? In such case what you commented at [1] should be revised to make clear this property would not affect the `/openapi` registration globally, correct?

[1]
https://github.com/wildfly/wildfly-proposals/pull/251#discussion_r337532401

Reply:
@fabiobrz After further deliberation, I suggest we add support for a `mp.openapi.extensions.disable` property.  This can be specified via a global config source (e.g. system property) to disable openapi globally, or via an application config source (e.g. /META-INF/microprofile-config.properties) to enable/disable on a per-application basis.